website/blog/2023-09-26-http-streaming-back-pressure-and-cancellation/index.md

```diff
@@ -0,0 +1,259 @@
+app.get('/api/server', async (req, res) => {
+  const response = await fetch('http://localhost:8000/api/llm');
+  const reader = response.body.getReader();
+
+  while (true) {
+		// Get data from stream when it available
+    const {done, value} = await reader.read();
+    if (done) {
+      break;
+    }
+    const chunk = new TextDecoder("utf-8").decode(value);
+    console.log(chunk); 
+  }
+})
+
+```
+
+
+But only using stream can not solve the back-pressure issue in the server side. If the stream between server and LLM is faster than the stream between server and client, the server would still save too much data in the memory till out-of-memory.
+
+![stream-flush](./stream-flush.png)
+
+
+That’s why we need to make use of `ReadableStream` and its method `pull` to
+manually fetch the data from the stream when needed.
+
+
+```JavaScript
+
+// Demo code for the server
+app.get('/api/server', async (req, res) => {
+	// 1. Read stream from LLM
+  const response = await fetch('http://localhost:8000/api/llm');
+  const reader = response.body.getReader();
+
+  // 2. Create a ReadableStream
+  const serverStream = new ReadableStream({
+    async pull(controller) {
+      const { value, done } = await reader.read();
+      if (done) {
+        controller.close();
+      } else {
+        controller.enqueue(value);
+      }
+    }
+  });
+  const serverReader = serverStream.getReader();
+
+	// 3. Custom fetching-data-from-llm logic based on the frontend needs
+  for (let i = 0; i < 3; i++) {
+		// Trigger the `pull` function and fetch the data from stream 
+    const { value } = await serverReader.read();
+    const chunk = new TextDecoder("utf-8").decode(value);
+    console.log(`read ${chunk}`); 
+ 
+    await sleep(1000);
+  }
+})
+
+```
+
+
+## What is cancellation?
+
+Now we know the backend is made up by the stream, and the stream usually is expensive and time-consuming. What if the client abort the in-flight request because of the network issue or other intended behaviors? That’s why we need to implement the cancellation to stop the stream on time in order to save the computer resource.
+
+![cancellation](./cancellation.png)
+
+## How to handle canellation?
+
+The core idea is straightforward: on the server side, we need to listen to the `close` event and check if the connection is still valid before pulling data from the LLM stream.
+
+```JavaScript
+app.get('/api/server', async (req, res) => {
+  let isConnectionClosed = false
+
+	// 1. Listen to the close event
+  req.on('close', () => {
+    console.log('close event')
+    isConnectionClosed = true
+  })
+
+  const response = await fetch('http://localhost:8000/api/llm');
+  const reader = response.body.getReader();
+
+  const serverStream = new ReadableStream({
+    async pull(controller) {
+      const { value, done } = await reader.read();
+      if (done) {
+        controller.close();
+      } else {
+        controller.enqueue(value);
+      }
+    }
+  });
```

Comment:
The code sample appears to have mixed `\t` and space indentation, as well as variation in the use of semicolons.  
It is recommand use Prettier to format the code sample.

Reply:
Thanks, didn't notice that when I was writing these on Notion at first